### PR TITLE
Update validation

### PR DESCRIPTION
**Pull request by paulci**
### Changes:
* Validate raw request data
* Update datetime objects to use utc

### CODE REVIEW

File: slingshotapi/auth/validation.py [hunk 1]
```diff
@@ -44,6 +41,37 @@ def _validate_hash(cls, secret_key, payload, str_timestamp):
...
+                cls._validate_hash(secret_key, str_payload, str_timestamp)
+                func()
+            except (
+                ManageValidation.HashCompareError,
+                ManageValidation.TimestampValidationError,
+                ManageValidation.RequestValidationError,
+                Authenticator.AuthenticationException
+            ) as e:
+                metrics.info('deployments.deploy.attempt 1 project={}'.format(project))
+                metrics.error('deployments.deploy.failed 1 project={} cause=payload-validation'.format(
```
Comment (balintzs) [inline slingshotapi/auth/validation.py]: These metrics do not belong here. Validation may not only be used for deployments, but other operations as well.

File: slingshotapi/auth/validation.py [hunk 2]
```diff
@@ -44,6 +41,37 @@ def _validate_hash(cls, secret_key, payload, str_timestamp):
...
+            print(flask.request.data)
+            project = payload_dict.get('project', 'unknown')
+            if not flask.current_app.config.get("ENFORCE_USE_API_ACCESS_KEYS"):
+                return func()
+            str_timestamp = flask.request.headers.get('X-SLINGSHOT-TIMESTAMP')
+            secret_key = Authenticator.get_secret_key()
+            try:
+                cls._validate_timestamp(str_timestamp)
+                cls._validate_hash(secret_key, str_payload, str_timestamp)
+                func()
```
Comment (balintzs) [inline slingshotapi/auth/validation.py]: You should return what the wrapped function returns.

File: slingshotapi/auth/validation.py [hunk 3]
```diff
@@ -44,6 +41,37 @@ def _validate_hash(cls, secret_key, payload, str_timestamp):
...
-        if (datetime.now() - timestamp).total_seconds() <= auth_threshold:
+        if (datetime.utcnow() - timestamp).total_seconds() <= auth_threshold:
             return True
         raise cls.TimestampValidationError("Timestamp is invalid")
+
+    @classmethod
+    def validate_request(cls, func):
+        exception_msg = 'Could not trigger deployment - {}'
+
+        def wrapper():
```
Comment (balintzs) [inline slingshotapi/auth/validation.py]: wrapper should take `*args` and `**kwargs` and pass the same int `func`

File: slingshotapi/blueprints/deployments.py [hunk 4]
```diff
@@ -11,6 +11,7 @@
 
 
 @deployments.route("/deployments", methods=['PUT'])
+@ManageValidation.validate_request
```
Comment (balintzs) [inline slingshotapi/blueprints/deployments.py]: Also remove validation exception handling from this function.

File: slingshotapi/deploy.py [hunk 5]
```diff
@@ -29,20 +28,9 @@ class UnsupportedVCSError(Exception):
 
     @classmethod
     def deploy(cls, deployment_data):
-        # Validate Payload
-        project = deployment_data.get('project', "unknown")
-        metrics.info('deployments.deploy.attempt 1 project={}'.format(project))
```
Comment (balintzs) [inline slingshotapi/deploy.py]: Do not remove this metric logging.

File: tests/test_validation.py [hunk 6]
```diff
@@ -35,38 +36,36 @@ def test_invalid_hash(flaskr, deployment_data):
...
-    with flaskr.test_request_context(headers=headers):
+    with flaskr.test_request_context(headers=headers, data=json.dumps(deployment_data)):
         flaskr.config['API_ACCESS_KEYS'] = dict(
             tool1="key1"
         )
         with flaskr.app_context():
-            with pytest.raises(ManageValidation.HashCompareError) as exec_info:
-                ManageValidation.validate_request(deployment_data)
-        assert str(exec_info.value) == "Hash is invalid"
+            exec_info = ManageValidation.validate_request(deployment_data)
```
Comment (balintzs) [inline tests/test_validation.py]: Confusing naming `exec_info` as well as bad argument. It should be something like `wrapped_function` taking a function as an argument.

File: slingshotapi/auth/validation.py [hunk 7]
```diff
@@ -44,6 +41,37 @@ def _validate_hash(cls, secret_key, payload, str_timestamp):
...
+            str_payload = flask.request.data.decode()
+            payload_dict = yaml.load(flask.request.data)
+            print(flask.request.data)
```
Comment (balintzs) [inline slingshotapi/auth/validation.py]: also remove rogue `print`

Comment (paulci) [inline slingshotapi/auth/validation.py]: Strange, must be an issue with the commit as i don't have this locally

File: tests/test_validation.py (hunk 6 above)
Comment (paulci) [inline tests/test_validation.py]: This is following the theme of other tests.  What is your reasoning behind this query?

File: slingshotapi/deploy.py (hunk 5 above)
Comment (paulci) [inline slingshotapi/deploy.py]: Mistake.  Agreed

File: slingshotapi/auth/validation.py (hunk 3 above)
Comment (paulci) [inline slingshotapi/auth/validation.py]: Ok.

File: slingshotapi/auth/validation.py (hunk 2 above)
Comment (paulci) [inline slingshotapi/auth/validation.py]: Ok.

File: slingshotapi/auth/validation.py [hunk 8]
```diff
@@ -44,6 +41,36 @@ def _validate_hash(cls, secret_key, payload, str_timestamp):
...
+
+    @classmethod
+    def validate_request(cls, metric_path):
+        def decorator(func):
+            exception_msg = 'Could not complete action - {}'
+
+            def wrapper(*args, **kwargs):
+                str_payload = flask.request.data.decode()
+                if not flask.current_app.config.get("ENFORCE_USE_API_ACCESS_KEYS"):
+                    return func()
```
Comment (balintzs) [inline slingshotapi/auth/validation.py]: pass `*args, **kwargs`

File: slingshotapi/auth/validation.py [hunk 9]
```diff
@@ -44,6 +41,36 @@ def _validate_hash(cls, secret_key, payload, str_timestamp):
...
+                str_timestamp = flask.request.headers.get('X-SLINGSHOT-TIMESTAMP')
+                secret_key = Authenticator.get_secret_key()
+                try:
+                    cls._validate_timestamp(str_timestamp)
+                    cls._validate_hash(secret_key, str_payload, str_timestamp)
+                    return func()
```
Comment (balintzs) [inline slingshotapi/auth/validation.py]: pass `*args, **kwargs`

File: slingshotapi/auth/validation.py [hunk 10]
```diff
@@ -44,6 +41,36 @@ def _validate_hash(cls, secret_key, payload, str_timestamp):
...
+                except (
+                    ManageValidation.HashCompareError,
+                    ManageValidation.TimestampValidationError,
+                    ManageValidation.RequestValidationError,
+                    Authenticator.AuthenticationException
+                ) as e:
+                    metrics.info('{}.attempt 1'.format(metric_path))
```
Comment (balintzs) [inline slingshotapi/auth/validation.py]: this should be the first line of the function

File: tests/test_validation.py [hunk 11]
```diff
@@ -76,11 +74,15 @@ def test_modified_timestamp(flaskr, deployment_data, sign_with_timestamp):
...
-            with pytest.raises(ManageValidation.HashCompareError) as exec_info:
-                ManageValidation.validate_request(deployment_data)
-        assert str(exec_info.value) == "Hash is invalid"
+            wrapped_function = validation_test_dummy()
+            assert wrapped_function[0] == "Could not complete action - Hash is invalid"
+
+
+@ManageValidation.validate_request(metric_path='deployments.deploy')
+def validation_test_dummy():
+    return
```
Comment (balintzs) [inline tests/test_validation.py]: return message and response code

File: tests/test_validation.py [hunk 12]
```diff
@@ -35,38 +35,36 @@ def test_invalid_hash(flaskr, deployment_data):
...
-    with flaskr.test_request_context(headers=headers):
+    with flaskr.test_request_context(headers=headers, data=json.dumps(deployment_data)):
         flaskr.config['API_ACCESS_KEYS'] = dict(
             tool1="key1"
         )
         with flaskr.app_context():
-            with pytest.raises(ManageValidation.HashCompareError) as exec_info:
-                ManageValidation.validate_request(deployment_data)
-        assert str(exec_info.value) == "Hash is invalid"
+            wrapped_function = validation_test_dummy()
```
Comment (balintzs) [inline tests/test_validation.py]: The return value of `validation_test_dummy` is not a function. `validation_test_dummy` itself is the wrapped function.

File: slingshotapi/auth/validation.py (hunk 10 above)
Comment (balintzs) [inline slingshotapi/auth/validation.py]: also, `attempt` is logged in the business logic as well. I would rename it to `{}.validation.attempt`, same for the failed validation and for that matter we could also add successful validation metric as well.

Comment (paulci) [inline slingshotapi/auth/validation.py]: I disagree.  Attempt is logged in the exception to ensure that it is not logged twice, in the event of success.  As we expect all attempts to be validated, then there is no need for the addtional logging

Comment (balintzs) [inline slingshotapi/auth/validation.py]: OK, fine. We can leave it like that for now.

File: tests/test_validation.py (hunk 12 above)
Comment (paulci) [inline tests/test_validation.py]: The return value is not relevant as the wrapped function will never be called.  These tests are anticipated exceptions.

Comment (balintzs) [inline tests/test_validation.py]: We discussed this. It should be tested in the following places:
* https://github.com/Skyscanner/slingshot-api/blob/yamlupdate/tests/test_validation.py#L10
* https://github.com/Skyscanner/slingshot-api/blob/yamlupdate/tests/test_validation.py#L28

**Comment by balintzs** [comment]
LGTM

**Pull request merged**